src/us/kbase/workspace/test/kbase/DelegatingTypeProviderTest.java

```diff
@@ -0,0 +1,487 @@
+	}
+	
+	// #### Build error tests ####
+	
+	@Test
+	public void getBuilderFail() throws Exception {
+		try {
+			DelegatingTypeProvider.getBuilder(null);
+			fail("expected exception");
+		} catch (Exception got) {
+			TestCommon.assertExceptionCorrect(got, new NullPointerException("client"));
+		}
+	}
+	
+	@Test
+	public void withTypeCacheTimeMSFail() throws Exception {
+		final WorkspaceClient c = mock(WorkspaceClient.class);
+		for (final int time: list(-1, -100, -1000000000)) {
+			try {
+				DelegatingTypeProvider.getBuilder(c).withTypeCacheTimeMS(time);
+				fail("expected exception");
+			} catch (Exception got) {
+				TestCommon.assertExceptionCorrect(got, new IllegalArgumentException(
+						"typeCacheTimeMS must be >= 0"));
+			}
+		}
+	}
+	
+	@Test
+	public void withTypeCacheMaxSizeFail() throws Exception {
+		final WorkspaceClient c = mock(WorkspaceClient.class);
+		for (final int size: list(-1, -100, -1000000000)) {
+			try {
+				DelegatingTypeProvider.getBuilder(c).withTypeCacheMaxSize(size);
+				fail("expected exception");
+			} catch (Exception got) {
+				TestCommon.assertExceptionCorrect(got, new IllegalArgumentException(
+						"typeCacheMaxSize must be >= 0"));
+			}
+		}
+	}
+	
+	@Test
+	public void withJsonSchemaCacheMaxSizeFail() throws Exception {
+		final WorkspaceClient c = mock(WorkspaceClient.class);
+		for (final int size: list(1024 * 1024 - 1, 1000000, 100, 0, -1, -100, -1000000000)) {
+			try {
+				DelegatingTypeProvider.getBuilder(c).withJsonSchemaCacheMaxSize(size);
+				fail("expected exception");
+			} catch (Exception got) {
+				TestCommon.assertExceptionCorrect(got, new IllegalArgumentException(
+						"jsonschemaCacheMaxSize must be >= 1048576"));
+			}
+		}
+	}
+	
+	// #### Type fetch error tests ####
+	
+	@Test
+	public void getJsonSchemaFailNoSuchModule() throws Exception {
+		// test with absolute & non-absolute types
+		for (final String type: list("Foo.Baz-32", "Foo.Baz-32.19")) {
+			final List<String> errors = list(
+					"Module doesn't exist: Foo",
+					"Module Foo was not initialized. For that you must request ownership of the "
+							+ "module, and your request must be approved.",
+					"Module wasn't uploaded: Foo"
+					);
+			for (final String err: errors) {
+				final TestMocks m = initMocks();
+				when(m.ws.getTypeInfo(type)).thenThrow(
+						new ServerException(err, 1, "somename", "stack\ntrace\nlline 1: i++"));
+				final Exception got = failGetJsonSchema(
+						m.p, TypeDefId.fromTypeString(type),
+						new NoSuchModuleException(err));
+				TestCommon.assertExceptionCorrect(
+						got.getCause(), new ServerException(err, 1, "somename"));
+			}
+		}
+	}
+	
+	@Test
+	public void getJsonSchemaForNonAbsoluteTypeFailNoSuchType() throws Exception {
+		// test with absolute & non-absolute types
+		for (final String type: list("Foo.Baz-32", "Foo.Baz-32.19")) {
+			final List<String> errors = list(
+					"This type wasn't released yet and you should be an owner to access "
```

Comment:
"hasn't been released yet" sounds more natural

Reply:
Theoretically I can, but I would much prefer not to since it comes from the type DB portion of the workspace which has pretty low test coverage, especially for unhappy path cases, and I'm not particularly familiar with the code. As such making changes there has a lot more risk associated with in than making changes in the workspace proper.

Reply:
A very reasonable approach.